### DISCUSSION TITLE
"pcks11-tool" and reader names greater than 64 characters

### DISCUSSION THREAD

**Discussion by srsross** [OpenSC/OpenSC#2776] [Q&A] [answered]
Is there a way to use `pkcs11-tool --slot-description READERNAME` (or some other set of options to `pkcs11-tool`) to pick a reader based on the reader's longer-than-64-character name?

Details follow...

Suppose that I have two smart card readers with long names, like the following
```
SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Interface] (12345678901234) 00 00
SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Interface] (ABCDEF12345678) 01 00
```
where the reader names are identical for the first 64 characters.  The names are different beginning with the USB serial numbers.

Only the first 64 characters of the reader name are significant, because that is the length of the `slotDescription` field in the CK_SLOT_INFO structure:
```
typedef struct CK_SLOT_INFO {
  CK_UTF8CHAR slotDescription[64];
  ...
} CK_SLOT_INFO;
```
So, for the above two reader names, the command `pkcs11-tool --slot-description READERNAME` fails to find the slot unless I first truncate the READERNAME to either of the following two arguments to `--slot-description`:
```
SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Int        <-- 61 characters long
SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Int...     <-- 64 characters long, ending with three dot characters.
```
But then the command may not pick the correct reader of the two different readers.  So, is there a work-around?

Thanks,
-- Steve Ross

**Comment by dengert** [comment]
This might help show where the reader name comes from:
https://ludovicrousseau.blogspot.com/2010/05/what-is-in-pcsc-reader-name.html

Note that the tokenInfo has token serial and token label (if available). So if you are only looking at slots with tokens, you could select the slot with the token with the label.

pkcs11-tool has these options to to select a slot:
      --slot <arg>              Specify the ID of the slot to use
      --slot-description <arg>  Specify the description of the slot to use
      --slot-index <arg>        Specify the index of the slot to use
      --object-index <arg>      Specify the index of the object to use
      --token-label <arg>       Specify the token label of the slot to use

Also note if running in a VM guest the reader name is derived from what the VM sees.
With VirtualBox with Windows 11 host and Ubuntu 22.4 as guest, The VM devices drop down devices menu show I have two `Yubico YubiKey OTP+FIDO+CCID [526]` device I can attach to the VM. These show up as:
```
./pkcs11-tool --list-slots
Available slots:
Slot 0 (0x0): Yubico YubiKey OTP+FIDO+CCID 00 00
  token label        : Doug_Y8
  token manufacturer : piv_II
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : 818cf26b716ae666
  pin min/max        : 4/8
Slot 1 (0x4): Yubico YubiKey OTP+FIDO+CCID 01 00
  token label        : deengert
  token manufacturer : piv_II
  token model        : PKCS#15 emulated
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : a84187a97da31764
  pin min/max        : 4/8
```


If using some program that will take a URI as defined in https://datatracker.ietf.org/doc/html/rfc7512 
and have p11-kit configured you can see what these look like are:
```
p11tool ----list-token-urls
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=d289a4f997d27222;token=dee-2
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=b77465002b1885f4;token=DEE-1
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=75b0ec1e3b974305;token=mytoken
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=ffa43bb8db818e0c;token=DEE-1
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=017c072ff770b055;token=dee-3
pkcs11:model=p11-kit-trust;manufacturer=PKCS%2311%20Kit;serial=1;token=System%20Trust
pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;serial=818cf26b716ae666;token=Doug_Y8
pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;serial=a84187a97da31764;token=deengert
```

**Comment by srsross** [comment]
@dengert, thanks for your response.

> This might help show where the reader name comes from:
https://ludovicrousseau.blogspot.com/2010/05/what-is-in-pcsc-reader-name.html

Thanks.  I'm familiar with this post.  I will need to refresh my memory, but I think it is possible to configure `pcscd` to only include parts of the reader name.  (But I may be wrong on that.)

> pkcs11-tool has these options to to select a slot:
...
--object-index Specify the index of the object to use
...

Is the argument for the `--object-index` option the value returned for the `Data object` (namely, `23703312`) in the following example output of `pkcs11-tool --list-objects` ?
```
...
Data object 23703312
  label:          'Card Capability Container' 
  application:    'Card Capability Container'
  app_id:         2.16.840.1.101.3.7.1.219.0
  flags:          <empty>
...
```
(These `Data object` values seem to change every time that I run the `--list-objects` command, so the values do not seem very useful.  Also, apparently they are not used in the current released version as per <https://github.com/OpenSC/OpenSC/issues/2697>.)

> If using some program that will take a URI as defined in https://datatracker.ietf.org/doc/html/rfc7512
and have p11-kit configured you can see what these look like are:
p11tool ----list-token-urls

Thank you for this suggestion; it jogged another thought for me...

Our system uses the (full) reader names to track tokens.  The `opensc-tool --list-readers` output lists both the full reader name and the corresponding "slot-index".  So, I think that I can first execute the `opensc-tool --list-readers` command, search the output for the reader name of interest, then note the corresponding "slot-index" and use that value in `pkcs11-tool --slot-index <INDEX>` instead of using the `--slot-description` option that I was trying to use in my previous post.  So, I sidestep the 64-character length limit  of the `slotDescription`.

**Reply by dengert** [reply, marked as answer]
I should probably not have listed  `--object-index` This and `--slot-index` can change from run to run or during as run. 

The "Data object" is really the  `CK_OBJECT_HANDLE`  which is really only usable within a PKCS11 application. 

Using the reader name is more like an index, that changes from time to time. It is not related to the token as it may have a USB index in it.  The is especially true with a smart card. 

The slot-index can also change between calls to C_GetSLotList if readers is added or removed or a card is removed from a reader or inserted in a reader. Or if C_GetSLotList is only looking for slots with tokens. 

You have referred to "Our system". Would you care to say more about it or what you are trying to do?
 
PKCS11 defines  `CK_UTF8CHAR slotDescription[64];`

What system produced such along name as "SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Interface] (12345678901234) 00 00"? 

This might also help:
https://ludovicrousseau.blogspot.com/2015/12/remove-andor-customize-pcsc-reader-names.html
Also see `pcscd` options  `--reader-name-no-serial`  and  `--reader-name-no-interface`  

The reader name is really out of control of OpenSC and PKCS11 defines `CK_UTF8CHAR slotDescription[64];`

**Comment by srsross** [comment]
Hi @dengert,

Thanks for another helpful response.
> This and --slot-index can change from run to run or during as run.

The output from `opensc-tool --list-readers` looks stable to me (barring removing or adding readers).  There would be a potential race condition between using that command to get an index for the reader and then providing the index to `pkcs11-tool --slot-index <INDEX>` if the user were to plug/unplug a reader during the time between the two calls (which you noted in your comment about `C_GetSlotList()`).

> The "Data object" is really the CK_OBJECT_HANDLE which is really only usable within a PKCS11 application.

Thank you.  I did not know that.

> You have referred to "Our system". Would you care to say more about it or what you are trying to do?

To add a few more details, our system/product keeps track of tokens via reader names.  In this particular instance, the software wants to read an X.509 certificate from the a particular token, but there is ambiguity regarding which certificate the software reads if two readers, of the same make/model, have a "long" (i.e., near the 64 character limit of `slotDescription`) `pcscd` friendly name.

> What system produced such along name as "SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Interface] (12345678901234) 00 00"?

The RHEL8 `pcsc-lite-ccid` package provides an `Info.plist` file with the "friendly name" of `SCM Microsystems Inc. SCR3310 USB Smart Card Reader` for the SCM SCR 3310 v.1 reader.  I am using two of those readers.

In addition, on RHEL8, the `pcscd` is configured, by default, to include the USB interface (namely, "CCID Interface") and the USB serial number ("12345678901234") by default.  So, that is the way that I get the greater-than-64-character reader name.

> This might also help:
https://ludovicrousseau.blogspot.com/2015/12/remove-andor-customize-pcsc-reader-names.html

Thanks for the reference.  I did not realize that the two options of `PCSCLITE_FILTER_IGNORE_READER_NAMES` and
`PCSCLITE_FILTER_EXTEND_READER_NAMES` existed.

> Also see pcscd options `--reader-name-no-serial` and `--reader-name-no-interface`

Ah, I think those are the two options that I need.  Without the USB interface and USB serial number, the `pcscd`-produced reader names have a much shorter length that puts the `pcscd`'s internal slot number inside the 64-character limit of `slotDescription`.  So, my software can distinguish between the two readers.

Thank you again.
-- Steve Ross

**Comment by srsross** [comment]
Hi @dengert and others,

A follow-up question on the options used to identify slots in the `pkcs11-tool` command, namely `--slot`, and `--slot-index`.....

The `pkcs11-tool --list-slots` command returns:
```
$ pkcs11-tool --list-slots
Available slots:
Slot 0 (0x0): SCM Microsystems Inc. SCR 3310 [CCID Interface] (533115492115...
  ...
Slot 1 (0x4): SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Int...
  ...
Slot 2 (0x8): SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Int...
...
```
By testing, I see that the "Slot _N_" number in the above output is used with the `--slot-index` option of  `pkcs11-tool`.  And also that the hexadecimal number "(0x _N_)" is used with the `--slot` option.  So far, so good.

Is there a relationship between the "Slot _N_" number and the output of the `opensc-tool --list-readers` command in the "Nr." column?  Are they the same value so that I should supply a reader number to the `--slot-index` option?
```
$ opensc-tool --list-readers
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             SCM Microsystems Inc. SCR 3310 [CCID Interface] (53311549211550) 00 00
1    Yes             SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Interface] (21120616202941) 01 00
2    Yes             SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Interface] (21120616202966) 02 00
```
(I note thatthe above `opensc-tool` command produces the same output as the `pcsc_scan -r` command below.
```
$ pcsc_scan -r
0: SCM Microsystems Inc. SCR 3310 [CCID Interface] (53311549211550) 00 00
1: SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Interface] (21120616202941) 01 00
2: SCM Microsystems Inc. SCR3310 USB Smart Card Reader [CCID Interface] (21120616202966) 02 00
```
)

Take a look at <https://ludovicrousseau.blogspot.com/2021/06/a-reader-for-96-smart-cards-sysmosimbank.html> which describes a a single device with 12 readers, where each reader has eight slots, for a total possible 96 smart cards.  Each of the 12 readers has its own USB serial number, which is common across all of its eight slots.

The interesting part is that you can see the mapping between the "reader numbers" in the left column as produced by the command `pcsc_scan -n` and the `pcscd`'s internal index number and its slot number in the right-most columns.  I am supposing that the command `pkcs11-tool --slot-index 95 ...` would be used to interact with the card in the highest numbered reader.

Thanks for any insight that you can offer,
-- Steve Ross

**Comment by dengert** [comment]
The problem is based on differences standards PCSC, PKCS11, CCID and others. For at least USB readers, the  `ifdManufacturerString`, `ifdProductString` are fetched from the device. These can be long as you have found out. 
PKCS11 was invented in 1995, USB was 1996, CCID version 1.0 was in 2001. PC/SC come from Windows NT and 95 and later from https://pcscworkgroup.com/specifications/  But windows does not use PKCS11, yet what we have today is PKCS11 modules using PCSC.  

`opensc-tool --list-readers`  and `pcsc_scan -r`  just list what PCSC returns and list the index of the entries returned. 

`pkcs11-tool --list-slots` lists the PKCS11 slots as returned by a call to `C_GetSlotList`, that more or less match the readers.  `pkcs11-tool` can uses PKCS11 modules from other packages. The  OpenSC  PKCS11 module  supports virtual slots to support smart cards that may have keys/objects protected by multiple PINS. PKCS11 `C_Login` really only supports the use of one user PIN per token. So for these types of tokens it will allocate a virtual slot of each. See https://github.com/OpenSC/OpenSC/blob/master/etc/opensc.conf.example.in#L1016-L1029 if you need to change it.

With PKCS11 each slot has a handle generated by the module. OpenSC uses sequential virtual handles. Thus when you see "Slot 1 (0x4): ..." you are seeing the slot index of the second slot with a reader, with the 0x4 being the 5th virtual slot.  Other PKCS11 modules may use different handles.  All you can really count on are the handles from the call to `C_GetSlotList`. `C_GetSlotList` can also be called to only return slots that have tokens too. The slot index is only valid between calls to `C_GetSlotList`

The use of "slot", "reader", "smart card" and "token" are often used but only valid within the context of the protocol/device being used.  "Token" some times mean a device which contains a reader and a smartcard. 

 Your example the device with 12 readers with 96 smart cards may or may not have a PKCS11 module that can support it. OpenSC does not, that I know of. I would not touch it. 

And if you plan to run on Windows see: https://learn.microsoft.com/en-us/troubleshoot/windows/win32/limitation-10-smart-card-readers This applies to any PKCS11 module that uses PCSC on windows.  

Many vendors provide a PKCS11 module for their device. Some contribute code to OpenSC to support their device.  If you need be able to use devices at the same time in the same application but not by single PKCS11 module See:  https://p11-glue.github.io/p11-glue/p11-kit.html which can load multiple PKCS11 modules and then present them to the application as a single PKCS11 module. P11-kit  maps the all the slots handles returned by these modules to new handles maintained by p11-kit.  Note that if p11-kit loads two modules that can both support the same token/smart card  you may see it twice by p11-kit. 
Example is OpenSC can support Yubikey token with PIV applet **OR** OpenPGP applet. You could also have Yubikey PIV PKCS11 module and/or some other OpenPGP PKCS11module. 

(The main reason to use PCSC is its ability to lock access to a reader/token for a set of transactions so other processes on the machine do not try and access the same token.) 

There are only two reasons I can think of where the application would need the reader name from PKCS11:
- You are going to insert uninitialized smart cards in a specific readers
- You are going to tell the user which reader to use. 
But in most cases any reader would do, as you can search for an initialized smart card by other means:

See https://www.rfc-editor.org/rfc/rfc7512  "The PKCS #11 URI Scheme".  and https://www.gnutls.org/manual/html_node/p11tool-Invocation.html `p11-tool --list-token-urls`
